src/propulsion/controller.cpp

```diff
@@ -0,0 +1,792 @@
+  }
+  log_.DBG1("MOTOR", enterOpMsgs_[2].logger_output, node_id_);
+  requestStateTransition(sdo_message_, kReadyToSwitchOn);
+
+  // Send enter operational message to transition to the Operation Enabled state
+  for (int i = 0; i < 8; i++) {
+    sdo_message_.data[i] = enterOpMsgs_[3].message_data[i];
+  }
+  log_.DBG1("MOTOR", enterOpMsgs_[3].logger_output, node_id_);
+  requestStateTransition(sdo_message_, kOperationEnabled);
+}
+
+void Controller::enterPreOperational()
+{
+  checkState();
+  if (state_ != kReadyToSwitchOn) {
+    // send shutdown command
+    if (sendControllerMessage(enterPreOpMsg_[0])) return;
+  }
+}
+
+void Controller::checkState()
+{
+  // Check Statusword in object dictionary
+  if (sendControllerMessage(checkStateMsg_[0])) return;
+}
+
+void Controller::sendTargetVelocity(int32_t target_velocity)
+{
+  // Send 32 bit integer in Little Edian bytes
+  for (int i = 0; i < 8; i++) {
+    sdo_message_.data[i] = sendTargetVelMsg[0].message_data[i];
+  }
+  sdo_message_.data[4] = target_velocity & 0xFF;
+  sdo_message_.data[5] = (target_velocity >> 8) & 0xFF;
+  sdo_message_.data[6] = (target_velocity >> 16) & 0xFF;
+  sdo_message_.data[7] = (target_velocity >> 24) & 0xFF;
+
+  log_.DBG2("MOTOR", sendTargetVelMsg[0].logger_output, node_id_, target_velocity);
+  sender.sendMessage(sdo_message_);
+  // TODO(Iain): ^ why is this one can.send instead of sendSdoMessage?
+}
+
+void Controller::sendTargetTorque(int16_t target_torque)
+{
+  // Send 16 bit integer in Little Edian bytes
+  for (int i = 0; i < 8; i++) {
+    sdo_message_.data[i] = sendTargetTorqMsg[0].message_data[i];
+  }
+  sdo_message_.data[4] = target_torque & 0xFF;
+  sdo_message_.data[5] = (target_torque >> 8) & 0xFF;
+
+  log_.DBG2("MOTOR", sendTargetTorqMsg[0].logger_output, node_id_, target_torque);
+  sendSdoMessage(sdo_message_);
+}
+
+void Controller::updateActualVelocity()
+{
+  // Check actual velocity in object dictionary
+  if (sendControllerMessage(updateActualVelMsg[0])) return;
+}
+
+void Controller::updateActualTorque()
+{
+  // Check actual torque in object dictionary
+  if (sendControllerMessage(updateActualTorqMsg[0])) return;
+}
+
+
+void Controller::quickStop()
+{
+  // Send quickStop command
+  if (sendControllerMessage(quickStopMsg[0])) return;
+}
+
+void Controller::healthCheck()
+{
+  // Check warning status & Check error status
+  for (int i = 0; i < 2; i++) {
+    if (sendControllerMessage(healthCheckMsgs[i])) return;
+  }
+}
+
+void Controller::updateMotorTemp()
+{
+  // Check motor temp in object dictionary
+  if (sendControllerMessage(updateMotorTempMsg[0])) return;
+}
+
+void Controller::updateControllerTemp()
+{
+  // Check controller temp in object dictionary
+  if (sendControllerMessage(updateContrTempMsg[0])) return;
+}
+
+void Controller::sendSdoMessage(utils::io::can::Frame& message)
+{
+  sdo_frame_recieved_ = false;
+  int8_t send_counter = 0;
+
+  for (send_counter = 0; send_counter < 3; send_counter++) {
```

Comment:
you can replace this entire for loop with this:

sender.sendMessage(message);
while(sender.getIsSending());

isSending will be set to false, when the processNewData function is called. Last years implementation did not have it set up this way, that's why they tried to send it 3 times.

You could add a timeout, like after 100ms it just says no response and then you throw a critical error. I don't know the response time, but 100ms should be high enough for now